docs/design.md

```diff
@@ -0,0 +1,156 @@
+important things.
+
+### Optional Dependencies
+
+In Helm there are two ways to handle dependencies.
+
+1. Explicitly declare other charts as dependencies. This will cause those charts
+   to be installed. When the templates are rendered they happen in the same
+   sandboxed environment.
+2. Install charts separately and use values to connect them to each other. For
+   example, install PostreSQL with one chart and then pass connection information
+   into another chart using a Secret or values that generate a Secret.
+
+Hypper intends to add a 3rd form of dependency via an optional dependency. For
+example, you are developing something that requires a prometheus CRD. You can
+depend on a chart that provides that CRD but does not provide prometheus. The
+CRD chart or your chart can declare an optional dependency on a chart that
+installs Prometheus. When using hypper to install your chart you will be asked
+about installing the optional dependencies. In development you can choose not
+to install the dependencies while in production you can choose to install them.
+
+The optional dependency handling will extend to all commands.
+
+Optional dependencies will be specified using Helm chart annotations.
+
+### Shared Dependencies
+
+Sometimes you want to have a shared system dependency. For example, you may want
+to have just a single Prometheus or Istio in a cluster. When you install an
+application that depends on this shared dependency you want to install it if not
+present and leverage the existing one if present.
+
+With Helm there is a separation between those who provide charts and those who
+consume them. Charts from various repositories can be installed in the same
+cluster but are not guaranteed to work together. That is left as an exercise for
+the chart authors.
+
+Hypper will make it easier to have a repository of charts that work together but
+can be installed separately.
+
+The additional metadata will be stored as Helm chart annotations.
+
+### OCI Storage
+
+OCI artifacts are a new way to store things. Helm supports storing Helm charts
+in OCI distributions as an experimental feature. Many container registries
+support storing non-images in them. The list includes Azure, AWS, Google Cloud,
+GitHub Container Registry, and Harbor. The one notable exception is Docker Hub.
+
+Kubernetes already needs to have a container registry. In air gapped, on-premise,
+and custom environments you need a registry to store your images. If that can be
+used to store charts instead of a separate system you have fewer system
+dependencies.
+
+Hypper will make OCI registries a primary element in the system.
+
+One notable disadvantage to using an OCI registry for storing charts is a lack of
+an `index.yaml` file. This file has metadata enabling some functionality that
+can't be duplicated with an OCI registry due to APIs not existing (e.g., search).
+To restore these features we are looking to put an index into an OCI registry.
```

Comment:
is it possible to expand on this? Does this mean pushing upstream to support having an index.yaml file or something completely different?

Reply:
> is it possible to expand on this? Does this mean pushing upstream to support having an index.yaml file or something completely different?

OCI registries, where you can push and pull containers from, implement the [distribution spec](https://github.com/opencontainers/distribution-spec). It has a catalog API that will list things you have access to. You need to page over it. Imagine doing that on docker hub? In practice it does not work. There is no search API and I don't expect one in the near future.

We are thinking of taking the `index.yaml` file for a Helm repo and putting it into an OCI registry like an image or Helm chart. Then you can add a Helm repo by telling Hypper to look at the OCI endpoint (e.g., `hypper repo add example oci://ghcr.io/mattfarina/examplerepo`).

Working with OCI registries this way is like working with object storage.

If you have some time take a look at `helm search`. We cannot do that with OCI, unfortunately.

When you have the metadata in an `index.yaml` file you can do a lot of resolution work without pulling down all the packages.

Reply:
Can't the metadata be added into the image itself?
I think it would be good not rely on an index.yaml that's published into a OCI image. While the image brings the immutability and the nice layers, I think metadata of the possible images containing the helm chart is having a better isolation, and we won't have the issue "my index.yaml container is outdated". Everything would be just auto wired. WDYT?

Reply:
The idea is to add an `index.yaml` as an object to the OCI registry, not to push an image with an `index.yaml` file to the registry.
Having an index in the registry means that one doesn't need to download and parse all the hypper objects of the registry to learn their dependency graph. This index.yaml object ideally would work for other object types in the OCI registry (container images, helm charts, etc), should they wish to use it. That would make it easier to upstream and have several OCI registry implementations understanding it.

Reply:
> Can't the metadata be added into the image itself?

A few things come to mind...

1. How would a search feature work? In an OCI registry of any size it will be almost impossible to perform a search by looking at each item in the registry, looking at metadata, and creating a response. How would search work?
2. Having an `index.yaml` file means you can take the one file and calculate the dependency tree without having to look at the other locations. All of the metadata is there. If this is cached locally you can do this without going over the network. This could all be handled by going over the network. It would just take more time and you would require network access to do it.

The biggest reason for an index file is in order to be able to display a list of charts and to search it. If not for an index, how would you display a lot of things for an end user to see and decide to install? How would you perform search?